File: packages/nats-client-ts/examples/dedup_lru.ts
/**
 * Cross-language dedup-LRU example (TS half).
 *
 * Per design Section 6 / P-DEDUP axis (P1-5 regression guard): the
 * in-process event_id LRU MUST evict by recency, not insertion order.
 * P1-5 was a TS regression where the LRU silently degraded to a Set
 * with FIFO eviction — every entry was kept until GC, then dropped in
 * batches that included recently-touched entries.
 *
 * Test pattern (capacity-agnostic — works for any LRU backed by
 * recency-aware eviction):
 *   1. Fill the LRU to capacity (N inserts).
 *   2. "Touch" id_target (one of the existing entries) by calling
 *      `has()` — a true LRU refreshes recency on read.
 *   3. Insert (N - 1) NEW entries. Each evicts the oldest. After all
 *      inserts, only id_target + the (N - 1) new entries fit.
 *   4. Replay id_target — `has()` MUST return true. Handler MUST NOT
 *      fire.
 *
 * Prints `{"handler_fired": <bool>}` to stdout. Expected value is
 * `false`. The orchestrator at
 * `tests/integration/nats-infra/cross-language-wire/orchestrator-dedup.py`
 * fails (exit 2) if any language reports `handler_fired: true`.
 */

import { EventIdLru } from "../src/consumers.js";

const CAPACITY = 1000;
const TARGET_INDEX = 500;
const targetId = `id-${TARGET_INDEX}`;

const lru = new EventIdLru();

// Step 1: fill to capacity.
for (let i = 0; i < CAPACITY; i++) {
  // Production semantic: handler fires when the id is novel; afterwards
  // we record it. We model that here directly.
  if (!lru.has(`id-${i}`)) lru.remember(`id-${i}`);
}

// Step 2: touch the target. has() returns true and refreshes recency.
const touched = lru.has(targetId);
if (!touched) {
  // The target should still be in the LRU at this point — if it's not,
  // step 1 didn't insert what we expected.
  throw new Error("dedup_lru: target evicted before touch — test setup bug");
}

// Step 3: insert (CAPACITY - 1) new entries. Each evicts the oldest.
// After all inserts, the LRU contains targetId + the new entries.
for (let i = 0; i < CAPACITY - 1; i++) {
  const id = `id-new-${i}`;
  if (!lru.has(id)) lru.remember(id);
}

// Step 4: replay the target. A true LRU still remembers it; handler
// must NOT fire. A Set-backed FIFO would have evicted it.
const handlerFired = !lru.has(targetId);

process.stdout.write(JSON.stringify({ handler_fired: handlerFired }) + "\n");

File: packages/nats-client-py/examples/dedup_lru.py
#!/usr/bin/env python3
"""Cross-language dedup-LRU example (Py half).

Per design Section 6 / P-DEDUP axis (P1-5 regression guard): the
in-process event_id LRU MUST evict by recency, not insertion order.

Test pattern (capacity-agnostic — works for any LRU backed by
recency-aware eviction):

  1. Fill the LRU to capacity (N inserts).
  2. "Touch" id_target (one of the existing entries) by calling
     ``has()`` — a true LRU refreshes recency on read.
  3. Insert (N - 1) NEW entries. Each evicts the oldest. After all
     inserts, only id_target + the (N - 1) new entries fit.
  4. Replay id_target — ``has()`` MUST return true. Handler MUST NOT
     fire.

Prints ``{"handler_fired": <bool>}`` to stdout. Expected value is
``false``. The orchestrator at
``tests/integration/nats-infra/cross-language-wire/orchestrator-dedup.py``
fails (exit 2) if any language reports ``handler_fired: true``.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

# Production LRU lives in the package's consumers module. Importing the
# private ``_EventIdLru`` directly (rather than a public re-export)
# matches the TS example which imports from `src/consumers.js` — both
# bypass the public surface so the test exercises the same primitive
# the production consume loop uses.
_PKG_SRC = Path(__file__).resolve().parent.parent / "src"
sys.path.insert(0, str(_PKG_SRC))

from klodi_nats_client.consumers import (  # noqa: E402
    DEDUP_LRU_SIZE,
    _EventIdLru,
)


def main() -> int:
    capacity = DEDUP_LRU_SIZE
    target_id = f"id-{capacity // 2}"

    lru = _EventIdLru()

    # Step 1: fill to capacity.
    for i in range(capacity):
        event_id = f"id-{i}"
        if not lru.has(event_id):
            lru.remember(event_id)

    # Step 2: touch the target — has() returns True and refreshes recency.
    if not lru.has(target_id):
        print(
            "py:error dedup_lru: target evicted before touch — test setup bug",
            file=sys.stderr,
        )
        return 1

    # Step 3: insert (capacity - 1) new entries. Each evicts the oldest;
    # the touched target stays alive because it was bumped to most-recent.
    for i in range(capacity - 1):
        event_id = f"id-new-{i}"
        if not lru.has(event_id):
            lru.remember(event_id)

    # Step 4: replay the target. A true LRU still remembers it.
    handler_fired = not lru.has(target_id)

    sys.stdout.write(
        json.dumps({"handler_fired": handler_fired}, separators=(",", ":")),
    )
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())
